pydantic/_internal/_generate_schema.py

```diff
@@ -283,6 +283,7 @@ def model_schema(self, cls: type[BaseModel]) -> core_schema.CoreSchema:
             config=core_config,
             post_init=model_post_init,
             metadata=metadata,
+            custom_init='__init__' in cls.__dict__,
```

Comment:
This way of setting the custom_init boolean doesn't preserve the setting for subclasses, which I think it should. Here's a failing test:
```python
def test_submodel_calls_custom_init():
    calls = []

    class InitModel(BaseModel):
        x: int
    
        def __init__(self, **kwargs):
            super().__init__(**kwargs)
            calls.append(kwargs)
    
    
    class SubInitModel(InitModel):
        pass
    
    
    SubInitModel.model_validate({'x': 1})
    assert calls
```
I think `custom_init=not (cls.__init__ is BaseModel.__init__)` might work better?

Reply:
Good catch, fixable I think.